x-pack/auditbeat/module/system/login/login.go

```diff
@@ -0,0 +1,193 @@
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/common/cfgwarn"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/metricbeat/mb"
+)
+
+const (
+	moduleName    = "system"
+	metricsetName = "login"
+
+	bucketName = "auditbeat.login.v1"
+
+	eventTypeEvent = "event"
+
+	// LoginRecordTypeBoot represents a system boot.
+	LoginRecordTypeBoot = "boot"
+	// LoginRecordTypeShutdown represents a system shutdown (halt or reboot).
+	LoginRecordTypeShutdown = "shutdown"
+	// LoginRecordTypeUserLogin represents a user login.
+	LoginRecordTypeUserLogin = "user_login"
+	// LoginRecordTypeUserLogout represents a user logout.
+	LoginRecordTypeUserLogout = "user_logout"
+)
+
+// LoginRecord represents a login record.
+type LoginRecord struct {
+	Utmp      Utmp
+	Type      string
+	PID       int
+	TTY       string
+	UID       int
+	Username  string
+	Hostname  string
+	IP        *net.IP
+	Timestamp time.Time
+	Origin    string
+}
+
+func (login LoginRecord) toMapStr() common.MapStr {
+	mapstr := common.MapStr{}
+
+	// Very useful for development
+	//mapstr.Put("utmp", fmt.Sprintf("%v", login.Utmp))
+
+	if login.TTY != "" {
+		mapstr.Put("tty", login.TTY)
+	}
+
+	return mapstr
+}
+
+func init() {
+	mb.Registry.MustAddMetricSet(moduleName, metricsetName, New,
+		mb.DefaultMetricSet(),
+	)
+}
+
+// MetricSet collects login records from /var/log/wtmp.
+type MetricSet struct {
+	mb.BaseMetricSet
+	config     Config
+	log        *logp.Logger
+	utmpReader *UtmpFileReader
+}
+
+// New constructs a new MetricSet.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	cfgwarn.Experimental("The %v/%v dataset is experimental", moduleName, metricsetName)
+
+	config := defaultConfig
+	if err := base.Module().UnpackConfig(&config); err != nil {
+		return nil, errors.Wrapf(err, "failed to unpack the %v/%v config", moduleName, metricsetName)
+	}
+
+	bucket, err := datastore.OpenBucket(bucketName)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to open persistent datastore")
+	}
+
+	ms := &MetricSet{
+		BaseMetricSet: base,
+		config:        config,
+		log:           logp.NewLogger(metricsetName),
+	}
+
+	ms.utmpReader, err = NewUtmpFileReader(ms.log, bucket, config.UtmpFilePattern)
+	if err != nil {
+		return nil, err
+	}
+
+	return ms, nil
+}
+
+// Close cleans up the MetricSet when it finishes.
+func (ms *MetricSet) Close() error {
+	return ms.utmpReader.Close()
+}
+
+// Fetch collects any new login records from /var/log/wtmp. It is invoked periodically.
+func (ms *MetricSet) Fetch(report mb.ReporterV2) {
+	loginRecords, err := ms.utmpReader.ReadNew()
+	if err != nil {
+		ms.log.Error(err)
+		report.Error(err)
+		return
+	}
+	ms.log.Debugf("%d new login records.", len(loginRecords))
+
+	for _, loginRecord := range loginRecords {
+		report.Event(ms.loginEvent(loginRecord))
+	}
+
+	// Save new state to disk
+	if len(loginRecords) > 0 {
+		err := ms.utmpReader.saveStateToDisk()
```

Comment:
Calling save to disk here can lead to data loss between restarts. In Winlogbeat/Filebeat we fill out the `Private` in the event and write the state update after an event has been ACKed by the output. The new registry file format will be more similar to a small in-memory key value store allowing you to reuse the pattern used in filebeat/winlogbeat.

Another potential issue might be the bucket name. The bucket name and therefore the actual storage for state is unique for this module, but not per instance. One must not have more then 2 instances collecting a file, or you will have 2 go-routines updating global state at the same time.

Reply:
Under which circumstances would the data loss happen? Even in case of a graceful shutdown? I suspect this can affect all metricsets (even the `auditd` module?), so at some point we might want to come up with a solution for all of Auditbeat (but maybe not now?).

Two Auditbeat instances using the same bucket concurrently would be a problem for sure. I would think that we should not allow two instances to use the same bucket, or even the same data path.

Reply:
I wouldn't block this metricset on the potential data loss. We can do better in the future, but I agree with @cwurm that from the user perspective it's the same guarantees as the other Auditbeat modules, so they won't expect more.

Reply:
> Under which circumstances would the data loss happen? Even in case of a graceful shutdown?

Yes.

> I wouldn't block this metricset on the potential data loss.

OK